src/ape_plugins/_cli.py

```diff
@@ -122,221 +155,110 @@ def _list(cli_ctx, display_all):
-                    any_install_failed = True
-            else:
-                cli_ctx.logger.warning(
-                    f"{module_name} is already installed. "
-                    f"Use the '--upgrade' option if you want to update '{module_name}'"
-                )
-
-        if not is_plugin_installed(module_name) and (
-            module_name in github_client.available_plugins
+        if upgrade:
+            cli_ctx.logger.info(f"Upgrading '{plugin.name}'...")
+            args.extend(("--upgrade", plugin.package_name))
+
+            version_before = plugin.current_version
+            result = subprocess.call(args)
+            failures_occurred = result_handler.handle_upgrade_result(result, version_before)
+
+        elif plugin.can_install and (
+            plugin.is_available
             or skip_confirmation
-            or click.confirm(f"Install unknown 3rd party plugin '{package_name}'?")
+            or click.confirm(f"Install unknown 3rd party plugin '{plugin.name}'?")
         ):
-            cli_ctx.logger.info(f"Installing {package_name}...")
+            cli_ctx.logger.info(f"Installing {plugin}...")
+            args.append(plugin.install_str)
+
             # NOTE: Be *extremely careful* with this command, as it modifies the user's
             #       installed packages, to potentially catastrophic results
             # NOTE: This is not abstracted into another function *on purpose*
-            if upgrade:
-                args.append("--upgrade")
-            args.append(package_name)
             result = subprocess.call(args)
-            plugin_got_installed = is_plugin_installed(module_name)
-            if result == 0 and plugin_got_installed:
-                cli_ctx.logger.success(f"Plugin '{module_name}' has been added.")
-            else:
-                cli_ctx.logger.error(f"Errors occurred when adding '{package_name}'.")
-                any_install_failed = True
-    if any_install_failed:
+            failures_occurred = not result_handler.handle_install_result(result)
+
+        else:
+            cli_ctx.logger.warning(
+                f"'{plugin.name}' is already installed. Did you mean to include '--upgrade'."
+            )
+
+    if failures_occurred:
         sys.exit(1)
 
 
-@cli.command(short_help="Uninstall all plugins in the local config file")
+@cli.command()
+@plugins_argument()
 @ape_cli_context()
 @skip_confirmation_option("Don't ask for confirmation to install the plugins")
-def uninstall(cli_ctx, skip_confirmation):
-    any_uninstall_failed = False
-    cwd = getcwd()
-    config_path = Path(cwd) / CONFIG_FILE_NAME
-    cli_ctx.logger.info(f"Uninstalling plugins from config file at {config_path}")
+def uninstall(cli_ctx, plugins, skip_confirmation):
+    """Uninstall plugins"""
 
-    plugins = config.get_config("plugins") or []
+    failures_occurred = False
+    did_warn_about_version = False
     for plugin in plugins:
```

Comment:
If `plugins` is empty, what happens?

Reply:
Similar thought, would be nice to have a `-a/--all` option to uninstall everything back to a clean state.

Reply:
Does `--all` uninstall _all_ plugins (including 3rd party)? Or does it just uninstall the ones from the config? Because just doing `ape plugins uninstall` with no arguments will already uninstall all the plugins in the config. I am a little nervous about offering a full plugin-nuke option. It seems like it is okay just to do `ape plugins uninstall plugin-not-in-config-a plugin-not-in-config-b` and maybe paired with a `ape plugins uninstall`

Reply:
Also, would `--all` uninstall plugins you have installed in editable mode?

Reply:
I think it should exit early and do nothing if you don't give any arguments of plugins to uninstall. `--all` could have it pick whatever is in the config file. Seems like if you're doing manual/editable installs, you probably would want to use `pip` directly.

Reply:
that makes it inconsistent with the behavior of `ape plugins install` though

Reply:
If `ape plugins install` uses the config when no plugins are given, I would expect `ape plugins uninstall` to do the same thing, no?

Reply:
I see how you did this. I am fine with the current behavior.